packages/backend/src/services/ProjectService/ProjectService.ts

```diff
@@ -1756,6 +1768,592 @@ export class ProjectService extends BaseService {
+            projectUuid: savedChartProjectUuid,
+            spaceUuid: savedChartSpaceUuid,
+            tableName: savedChartTableName,
+            metricQuery,
+        } = await this.savedChartModel.get(chartUuid, versionUuid);
+
+        // Check chart belongs to project
+        if (savedChartProjectUuid !== projectUuid) {
+            throw new ForbiddenError('Chart does not belong to project');
+        }
+
+        const [space, explore] = await Promise.all([
+            this.spaceModel.getSpaceSummary(savedChartSpaceUuid),
+            this.getExplore(
+                user,
+                projectUuid,
+                savedChartTableName,
+                savedChartOrganizationUuid,
+            ),
+        ]);
+
+        const access = await this.spaceModel.getUserSpaceAccess(
+            user.userUuid,
+            space.uuid,
+        );
+
+        if (
+            user.ability.cannot(
+                'view',
+                subject('SavedChart', {
+                    savedChartOrganizationUuid,
+                    projectUuid,
+                    isPrivate: space.isPrivate,
+                    access,
+                }),
+            ) ||
+            user.ability.cannot(
+                'view',
+                subject('Project', {
+                    savedChartOrganizationUuid,
+                    projectUuid,
+                }),
+            )
+        ) {
+            throw new ForbiddenError();
+        }
+
+        await this.analyticsModel.addChartViewEvent(
+            savedChartUuid,
+            user.userUuid,
+        );
+
+        const queryTags: RunQueryTags = {
+            organization_uuid: savedChartOrganizationUuid,
+            project_uuid: projectUuid,
+            user_uuid: user.userUuid,
+            chart_uuid: chartUuid,
+            explore_name: savedChartTableName,
+            query_context: context,
+        };
+
+        return this.runPaginatedQuery(
+            {
+                user,
+                projectUuid,
+                exploreName: savedChartTableName,
+                page,
+                pageSize,
+                context,
+                queryTags,
+                invalidateCache: false,
+                metricQuery,
+                csvLimit: undefined,
+            },
+            formatRow,
+        );
+    }
+
+    async runPaginatedDashboardChartQuery({
+        user,
+        projectUuid,
+        chartUuid,
+        dashboardUuid,
+        dashboardFilters,
+        dashboardSorts,
+        granularity,
+        autoRefresh,
+        page,
+        pageSize,
+        context = QueryExecutionContext.DASHBOARD,
+    }: PaginateDashboardChartArgs): Promise<ApiPaginatedQueryResults> {
+        if (!isUserWithOrg(user)) {
+            throw new ForbiddenError('User is not part of an organization');
+        }
+
+        const savedChart = await this.savedChartModel.get(chartUuid);
+        const { organizationUuid, projectUuid: dashboardProjectUuid } =
```

Comment:
chartProjectUuid rather than dashboardProjectUuid ?

Reply:
Or do we want to get both chart and dashboard and compare both project uuids with the one from the URL?